clippy_lints/src/casts/mod.rs

```diff
@@ -596,6 +597,32 @@ declare_clippy_lint! {
     "casting a slice created from a pointer and length to a slice pointer"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the result of a `&self`-taking `as_ptr` being cast to a mutable pointer
+    ///
+    /// ### Why is this bad?
+    /// Since `as_ptr` took a `&self`, the pointer won't have write permissions, making it
```

Comment:
_Won't_ is a bit too strong. This is not a dangerous pattern if the pointer points into an `UnsafeCell`.

I don't know how technically correct or detailed clippy docs are expected to be, but this should not be so absolute.

Reply:
I've added a `unless interior mutability is used` which makes this correct, but it doesn't sound great in my opinion.

Reply:
Perhaps
```
    /// Since `as_ptr` takes a `&self`, the pointer won't have write permissions unless interior
    /// mutability is used. If this cast to `*mut` exists to enable mutation through this pointer,
    /// that mutation is likely to be UB.
```